### PR TITLE
*: fix staled table cache usage inside a transaction

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Problem Summary:

A subtask of  https://github.com/pingcap/tidb/issues/25293

In session1:

```
begin;
select * from t;   // use table cache data
```

then in session2:

```
insert into t values (xxx);  // change the table cache data
select * from t;           // and reload the cache
```

in session1:

```
select * from t;       
// read again, although the cache exist, and cache.Lease > txn.StartTS()
//  it should not use the cache
```

### What is changed and how it works?

- Add a `Start` field to table cache, txn `TS` within `[Start, Lease)` can use the cache data
- Store table cache in plan, once the plan is decided, the execute result is decided

`begin; sleep(1h)...`, then do some write operation, and reload the cache... 

In this case, comparing the `txn.StartTS()` with the  `[Start, Lease)` of the cache will fail, because `txn.StartTS() < Start`, the cache is not used.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: table/tables/cache.go [hunk 1]
```diff
@@ -88,42 +95,45 @@ func NewCachedTable(tbl *TableCommon) (table.Table, error) {
...
 
-func (c *cachedTable) loadDataFromOriginalTable(store kv.Storage, lease uint64) (kv.MemBuffer, error) {
+func (c *cachedTable) loadDataFromOriginalTable(store kv.Storage, lease uint64) (kv.MemBuffer, uint64, error) {
 	buffer, err := newMemBuffer(store)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
+	var startTS uint64
 	err = kv.RunInNewTxn(context.Background(), store, false, func(ctx context.Context, txn kv.Transaction) error {
```
Comment (JayL-zxl) [inline table/tables/cache.go:104]: retryable should be true?

Comment (tiancaiamao) [reply table/tables/cache.go]: done

File: planner/core/stats.go [hunk 2]
```diff
@@ -431,8 +431,7 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo, selfSchema *
 			}
 		}
 	}
-	cond, _ := ds.ctx.GetSessionVars().StmtCtx.GetCacheTable(ds.tableInfo.ID)
-	if isPossibleIdxMerge && sessionAndStmtPermission && needConsiderIndexMerge && isReadOnlyTxn && ds.tableInfo.TempTableType != model.TempTableLocal && !cond {
+	if isPossibleIdxMerge && sessionAndStmtPermission && needConsiderIndexMerge && isReadOnlyTxn && ds.tableInfo.TempTableType != model.TempTableLocal {
```
Comment (lcwangchao) [inline planner/core/stats.go:434]: Why remove the if here? I think union_scan does not support index merge currently.

File: table/tables/cache_test.go [hunk 3]
```diff
@@ -236,3 +243,47 @@ func TestCacheTableComplexRead(t *testing.T) {
...
+			break
+		}
+	}
+	require.True(t, cacheUsed)
+
+	// Another session change the data and make the cache unavailable.
+	tk2.MustExec("update aba set v = 2")
+
+	// And then make the cache available again.
+	for {
```
Comment (lcwangchao) [inline table/tables/cache_test.go:279]: Better to have a timeout here.

File: planner/core/stats.go (hunk 2 above)
Comment (tiancaiamao) [reply planner/core/stats.go:434]: > union_scan does not support index merge currently.

Oh, I don't know that.... that's why we disable it in the past?

File: executor/distsql.go [hunk 4]
```diff
@@ -287,7 +281,7 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange)
 	// In a test case IndexReaderExecutor is mocked and e.table is nil.
 	// Cache table is similar with temporary table, if it satisfies the read condition.
 	// Avoid sending distsql request to TIKV.
-	if e.table != nil && e.table.Meta().TempTableType != model.TempTableNone || isReadFromCache(e.table, e.ctx.GetSessionVars().StmtCtx) {
```
Comment (JayL-zxl) [inline executor/distsql.go:290]: why delete this line ?

File: planner/core/stats.go [hunk 5]
```diff
@@ -431,8 +431,8 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo, selfSchema *
 			}
 		}
 	}
-	cond, _ := ds.ctx.GetSessionVars().StmtCtx.GetCacheTable(ds.tableInfo.ID)
-	if isPossibleIdxMerge && sessionAndStmtPermission && needConsiderIndexMerge && isReadOnlyTxn && ds.tableInfo.TempTableType != model.TempTableLocal && !cond {
+	readFromTableCache := ds.ctx.GetSessionVars().StmtCtx.ReadFromTableCache
```
Comment (lcwangchao) [inline planner/core/stats.go:434]: if a statement involves two tables but only one of it is using table cache. Will index merge still be disabled for  another table?

**Review by lcwangchao — APPROVED** [review]

File: planner/core/stats.go (hunk 5 above)
Comment (lcwangchao) [reply planner/core/stats.go:434]: https://github.com/pingcap/tidb/issues/29682

File: executor/distsql.go [hunk 6]
```diff
@@ -287,7 +281,7 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange)
...
+	if e.table != nil && e.table.Meta().TempTableType != model.TempTableNone {
```
Comment (tangenta) [inline executor/distsql.go:284]: Please also update the comments.

File: executor/distsql.go [hunk 7]
```diff
@@ -412,7 +406,7 @@ func (e *IndexLookUpExecutor) Open(ctx context.Context) error {
 	}
 
 	// Treat temporary table as dummy table, avoid sending distsql request to TiKV.	Cache table is similar with temporary table.
-	if e.table.Meta().TempTableType != model.TempTableNone || isReadFromCache(e.table, e.ctx.GetSessionVars().StmtCtx) {
+	if e.table.Meta().TempTableType != model.TempTableNone {
```
Comment (tangenta) [inline executor/distsql.go:409]: Please also update the comments.

File: table/table.go [hunk 8]
```diff
@@ -252,17 +252,10 @@ var MockTableFromMeta func(tableInfo *model.TableInfo) Table
 type CachedTable interface {
 	Table
 
-	// TryGetMemcache Check if the cache table is readable, if it is readable,
-	// Return the pointer to the MemBuffer and true otherwise return nil and false
-	TryGetMemcache(ts uint64) (kv.MemBuffer, bool)
+	// TryReadFromCache Check if the cache table is readable
```
Comment (tangenta) [inline table/table.go:255]:
```suggestion
	// TryReadFromCache checks if the cache table is readable.
```

File: executor/distsql.go (hunk 6 above)
Comment (tiancaiamao) [reply executor/distsql.go]: Done

**Review by tangenta — APPROVED** [review]

**Pull request merged**